### PR TITLE
Add projects for supporting source generator testing

### PR DESCRIPTION
**Pull request by sharwell**

### CODE REVIEW

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.SourceGenerators.Testing/SourceGeneratorTest`1.cs [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+            var fixedState = rawFixedState.WithProcessedMarkup(MarkupOptions, defaultDiagnostic, supportedDiagnostics, fixableDiagnostics, DefaultFilePath);
+
+            await VerifyDiagnosticsAsync(new EvaluatedProjectState(testState, ReferenceAssemblies), testState.AdditionalProjects.Values.Select(additionalProject => new EvaluatedProjectState(additionalProject, ReferenceAssemblies)).ToImmutableArray(), testState.ExpectedDiagnostics.ToArray(), Verify.PushContext("Diagnostics of test state"), cancellationToken).ConfigureAwait(false);
+            await VerifyDiagnosticsAsync(new EvaluatedProjectState(fixedState, ReferenceAssemblies), fixedState.AdditionalProjects.Values.Select(additionalProject => new EvaluatedProjectState(additionalProject, ReferenceAssemblies)).ToImmutableArray(), fixedState.ExpectedDiagnostics.ToArray(), Verify.PushContext("Diagnostics of fixed state"), cancellationToken).ConfigureAwait(false);
+            await VerifySourceGeneratorAsync(testState, fixedState, Verify, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.SourceGenerators.Testing/SourceGeneratorTest`1.cs]: This is now fixed

File: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.SourceGenerators.Testing/SourceGeneratorVerifier`3.cs [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+        /// <summary>
+        /// Verifies the application of the source generator produces the expected result.
+        /// </summary>
+        /// <param name="source">The source text to test, which may include markup syntax.</param>
+        /// <param name="expected">The expected diagnostics. These diagnostics are in addition to any diagnostics
+        /// defined in markup.</param>
+        /// <param name="fixedSource">The expected fixed source text. Any remaining diagnostics are defined in markup.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        public static Task VerifySourceGeneratorAsync(string source, DiagnosticResult[] expected, string fixedSource)
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.SourceGenerators.Testing/SourceGeneratorVerifier`3.cs]: ➡️ For now I've just removed these.

**Comment by sharwell** [comment]
@chsienki @jmarolf for reviews

**Comment by jmarolf** [comment]
>     public sealed class EvaluatedProjectState

Curious if this could be a record

---
Refers to: src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Model/EvaluatedProjectState.cs:13 in 90c7987. [](commit_id = 90c7987a5d1ac7fd329d8171fec53aae1b021926, deletion_comment = False)

**Review by jmarolf — APPROVED** [review]
:shipit:

**Pull request merged**